src/common/snippets/src/lowered/pass/define_buffer_clusters.cpp

```diff
@@ -0,0 +1,340 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "snippets/lowered/pass/define_buffer_clusters.hpp"
+
+#include "snippets/lowered/pass/identify_buffers.hpp"
+#include "snippets/pass/tokenization.hpp"
+#include "snippets/itt.hpp"
+
+namespace ov {
+namespace snippets {
+namespace lowered {
+namespace pass {
+
+using BufferCluster = AllocateBufferMemory::BufferCluster;
+using BufferClusters = AllocateBufferMemory::BufferClusters;
+
+BufferClusters::iterator DefineBufferClusters::find_cluster_by_expr(const ExpressionPtr& target) {
+    return std::find_if(m_clusters.begin(), m_clusters.end(),
+                        [&target](const AllocateBufferMemory::BufferCluster& cluster) { return cluster.count(target) > 0; });
+}
+
+bool DefineBufferClusters::is_direct_buffer(const ExpressionPtr& buffer_expr, const ExpressionPtr& target_expr) const {
+    const auto buffer = ov::as_type_ptr<op::Buffer>(buffer_expr->get_node());
+    return buffer && buffer_expr->get_loop_ids() == target_expr->get_loop_ids();
+}
+
+void DefineBufferClusters::create_new_cluster(const ExpressionPtr& buffer_expr) {
+    const auto cluster_it = find_cluster_by_expr(buffer_expr);
+    // If Buffer is missed in clusters, create new cluster with the single Buffer node inside
+    if (cluster_it == m_clusters.cend()) {
+        m_clusters.push_back(BufferCluster{buffer_expr});
+    }
+}
+
+size_t DefineBufferClusters::get_cluster_buffer_id(const AllocateBufferMemory::BufferCluster& cluster) const {
+    OPENVINO_ASSERT(!cluster.empty(), "Buffer cluster is empty!");
+    const auto id = (ov::as_type_ptr<op::Buffer>(cluster.cbegin()->get()->get_node()))->get_id();
+    if (std::all_of(cluster.cbegin(), cluster.cend(),
+                    [&id](const ExpressionPtr& expr) { return (ov::as_type_ptr<op::Buffer>(expr->get_node()))->get_id() == id; })) {
+        return id;
+    }
+    return SIZE_MAX;
+}
+
+void DefineBufferClusters::get_io_buffers(BufferPorts& input_buffers, BufferPorts& output_buffers, const ExpressionPtr& loop_expr) const {
+    input_buffers.clear();
+    output_buffers.clear();
+
+    const auto loop_end = ov::as_type_ptr<op::LoopEnd>(loop_expr->get_node());
+    const auto in_count = loop_end->get_input_num();
+    const auto out_count = loop_end->get_output_num();
+    const auto connectors = loop_expr->get_input_port_connectors();
+
+    // Input Buffers
```

Comment:
Input/output buffers filling logic do not intersect. Maybe we can separate this method on `get_input_buffers` and `get_output_buffers`? In this case we can return `BufferPorts` from these functions and not pass them as parameters. As for me, it looks a bit better in terms of readability

Reply:
Good idea, done! Thanks a lot!